Treat performance.mark like console.timeStamp in Web Inspector's timeline

rniwa opened the pull request:
#### 4d52235be538b0fb8ab33a0d80f1d7b18799f8df
<pre>
Treat performance.mark like console.timeStamp in Web Inspector&apos;s timeline
<a href="https://bugs.webkit.org/show_bug.cgi?id=258385">https://bugs.webkit.org/show_bug.cgi?id=258385</a>

Reviewed by Patrick Angle.

Treat performance.mark like console.timeStamp and generate a timeline marker (vertical line).
This will help investigate various aspects of the Speedometer 3 benchmark.

Since performance.now can be called at anytime even when inspector&apos;s stopwatch is paused,
this patch also introduces Stopwatch::fromMonotonicTime to convert monotonic time to time
used by the stopwatch.

Also expose markers via a public accessor in TimelineRecording.

* LayoutTests/inspector/timeline/resources/timeline-event-utilities.js:
(TestPage.registerInitializer.InspectorTest.TimelineEvent.captureTimelineWithMarker): Added.
* LayoutTests/inspector/timeline/timeline-event-performance-mark-expected.txt: Added.
* LayoutTests/inspector/timeline/timeline-event-performance-mark.html: Added.

* Source/WTF/wtf/Stopwatch.h:
(WTF::Stopwatch::stop):
(WTF::Stopwatch::fromMonotonicTime): Added.

* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::performanceMarkImpl): Added.

* Source/WebCore/inspector/InspectorInstrumentation.h: Added.
(WebCore::InspectorInstrumentation::performanceMark):

* Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
(WebCore::InspectorTimelineAgent::timestampFromMonotonicTime): Added.
(WebCore::InspectorTimelineAgent::didPerformanceMark): Added.
(WebCore::InspectorTimelineAgent::appendRecord): Now optionally takes startTime.

* Source/WebCore/inspector/agents/InspectorTimelineAgent.h:

* Source/WebCore/page/Performance.cpp:
(WebCore::Performance::monotonicTimeFromRelativeTime const): Added.

* Source/WebCore/page/Performance.h:

* Source/WebCore/page/PerformanceMarkOptions.h:

* Source/WebCore/page/PerformanceUserTiming.cpp:
(WebCore::PerformanceUserTiming::mark):

* Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
(WI.TimelineRecording):
(WI.TimelineRecording.prototype.exportData):
(WI.TimelineRecording.prototype.get markers):
(WI.TimelineRecording.prototype.reset):
(WI.TimelineRecording.prototype.addEventMarker):

Canonical link: <a href="https://commits.webkit.org/265491@main">https://commits.webkit.org/265491@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9339dba4262de4896c112fa72d6f6d7675917aad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10796 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11016 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11317 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12446 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10351 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10811 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13390 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10993 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13255 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10956 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11866 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9084 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12849 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9160 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9741 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16997 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/9147 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10231 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9893 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13143 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/10245 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10366 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8451 "14 flakes 2 failures") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10909 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9524 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9632 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2975 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13797 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/11214 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1237 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10226 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2739 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

JosephPecoraro commented on LayoutTests/inspector/timeline/timeline-event-performance-mark.html:
```diff
@@ -0,0 +1,96 @@
+<script>
+
+TestPage.registerInitializer(() => {
+    InspectorTest.TimelineEvent.captureTimelineWithMarker = function({expression, markerType, timelineType}) {
```
You might as well put this in `timeline-event-utilities.js` to be shared with other tests if needed. I could see a test being written for the Load event marker.

rniwa replied:
Sure, will do.

JosephPecoraro commented on LayoutTests/inspector/timeline/timeline-event-performance-mark.html:
```diff
@@ -0,0 +1,96 @@
+
+            let recording = WI.timelineManager.activeRecording;
+
+            InspectorTest.expectEqual(recording._exportDataMarkers.length, 2);
```
We should probably rename `_exportDataMarkers` to `_markers` and add a public accessor for it (`get markers() { return _markers; }`) in `WI.TimelineRecording`. That would be a cleaner interface than reaching into its private properties.

JosephPecoraro commented on Source/WTF/wtf/Stopwatch.h:
```diff
@@ -90,6 +96,23 @@ inline Seconds Stopwatch::elapsedTimeSince(MonotonicTime timeStamp) const
     return m_elapsedTime + (timeStamp - m_lastStartTime);
 }
 
+inline std::optional<Seconds> Stopwatch::fromMonotonicTime(MonotonicTime timeStamp) const
```
I'm surprised there aren't `TestWebKitAPI` tests for Stopwatch. Otherwise I'd have expected some basic tests for this. If you're motivated enough to add them that would be awesome, but probably not a requirement =)

dcrousso commented on Source/WebCore/inspector/InspectorInstrumentation.h:
```diff
@@ -1677,6 +1683,18 @@ inline void InspectorInstrumentation::consoleStopRecordingCanvas(CanvasRendering
+
+inline void InspectorInstrumentation::performanceMark(WorkerOrWorkletGlobalScope& globalScope, const String& label, std::optional<MonotonicTime> startTime)
+{
+    performanceMarkImpl(instrumentingAgents(globalScope), label, WTFMove(startTime));
```
can we add a `FAST_RETURN_IF_NO_FRONTENDS(void());` here too?

also, i think you can drop the different overrides for `LocalFrame` and `WorkerOrWorkletGlobalScope` since there's a `InspectorInstrumentation::instrumentingAgents` that takes a `ScriptExecutionContext` and knows how to do the rest

rniwa replied:
Hm... we still have to pass the frame in InspectorTimelineAgent::appendRecord though but we can use a single function indeed.

dcrousso commented on Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
```diff
@@ -481,6 +489,17 @@ void InspectorTimelineAgent::timeEnd(LocalFrame& frame, const String& message)
+        if (!timestamp)
+            return; // Stopwatch wasn't running at the time of timeInMonotonicTime.
+    }
+    appendRecord(TimelineRecordFactory::createTimeStampData(label), TimelineRecordType::TimeStamp, true, frame, timestamp);
```
is the only reason that we have to pass along the `MonotonicTime` because `performance.mark(...)` allows for a custom `startTime`?  i ask because we don't do that for any of the similar things like `console.timeStamp()`, instead just using `InspectorTimelineAgent::timestamp` to grab the current time

IIRC the only time that we ever pause the `Stopwatch` is when the debugger is paused, and the Web Inspector frontend should prevent that from happening when Timelines stuff is happening (and vice versa), so do we actually even need to worry about whether the `Stopwatch` was running?  can you provide an example of where that'd be a problem?

rniwa replied:
Right, this is to support a custom startTime. This comes into play when we can pause the inspector's timeline recording & restart the recording again as far as I could tell.

dcrousso replied:
i dont think pausing and resuming a timeline recording affects the `Stopwatch`.  IIRC the only thing that affects it is the `Debugger`.

is there maybe some sort of odd scenario we're trying to support where the given `startTime` is somehow derived when the `Debugger` is paused?  if not, i dont think we actually need to worry about whether the `Stopwatch` was running or not

dcrousso commented on Source/WebCore/page/PerformanceUserTiming.cpp:
```diff
@@ -91,7 +93,17 @@ static void addPerformanceEntry(PerformanceEntryMap& map, const String& name, Pe
+        timestamp = m_performance.monotonicTimeFromRelativeTime(*markOptions->startTime);
+    if (auto* globalScope = dynamicDowncast<WorkerOrWorkletGlobalScope>(context.get()))
+        InspectorInstrumentation::performanceMark(*globalScope, markName, timestamp);
+    else if (auto* document = dynamicDowncast<Document>(context.get()); document && document->frame())
```
why not just use the `Document`?  `InspectorInstrumentation` can definitely handle that

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
```diff
@@ -154,6 +154,8 @@ WI.TimelineRecording = class TimelineRecording extends WI.Object
     get startTime() { return this._startTime; }
     get endTime() { return this._endTime; }
 
+    get markers() { return this._markers; }
```
to take what @JosephPecoraro said a step further, i actually think this should be kept as `_exportDataMarkers` and instead of this like, add something like this at the end of this file
```
// Testing

get markersForTesting() { return this._exportDataMarkers; }
```
since ultimately this isn't something that's used in actual production (and we'd probably want to keep it that way)

patrickangle reviewed (approved):
lgtm

rniwa committed:
Treat performance.mark like console.timeStamp in Web Inspector's timeline
https://bugs.webkit.org/show_bug.cgi?id=258385

Reviewed by Patrick Angle.

Treat performance.mark like console.timeStamp and generate a timeline marker (vertical line).
This will help investigate various aspects of the Speedometer 3 benchmark.

Since performance.now can be called at anytime even when inspector's stopwatch is paused,
this patch also introduces Stopwatch::fromMonotonicTime to convert monotonic time to time
used by the stopwatch.

Also expose markers via a public accessor in TimelineRecording.

* LayoutTests/inspector/timeline/resources/timeline-event-utilities.js:
(TestPage.registerInitializer.InspectorTest.TimelineEvent.captureTimelineWithMarker): Added.
* LayoutTests/inspector/timeline/timeline-event-performance-mark-expected.txt: Added.
* LayoutTests/inspector/timeline/timeline-event-performance-mark.html: Added.

* Source/WTF/wtf/Stopwatch.h:
(WTF::Stopwatch::stop):
(WTF::Stopwatch::fromMonotonicTime): Added.

* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::performanceMarkImpl): Added.

* Source/WebCore/inspector/InspectorInstrumentation.h: Added.
(WebCore::InspectorInstrumentation::performanceMark):

* Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
(WebCore::InspectorTimelineAgent::timestampFromMonotonicTime): Added.
(WebCore::InspectorTimelineAgent::didPerformanceMark): Added.
(WebCore::InspectorTimelineAgent::appendRecord): Now optionally takes startTime.

* Source/WebCore/inspector/agents/InspectorTimelineAgent.h:

* Source/WebCore/page/Performance.cpp:
(WebCore::Performance::monotonicTimeFromRelativeTime const): Added.

* Source/WebCore/page/Performance.h:

* Source/WebCore/page/PerformanceMarkOptions.h:

* Source/WebCore/page/PerformanceUserTiming.cpp:
(WebCore::PerformanceUserTiming::mark):

* Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
(WI.TimelineRecording):
(WI.TimelineRecording.prototype.exportData):
(WI.TimelineRecording.prototype.get markers):
(WI.TimelineRecording.prototype.reset):
(WI.TimelineRecording.prototype.addEventMarker):

Canonical link: https://commits.webkit.org/265491@main
